src/Orleans/IDs/Interner.cs

```diff
@@ -96,31 +96,34 @@ public Interner(int initialSize, TimeSpan cleanupFreq)
         /// <param name="key">key to find</param>
         /// <param name="creatorFunc">function to create new object and store for this key if no cached copy exists</param>
         /// <returns>Object with specified key - either previous cached copy or newly created</returns>
-        public T FindOrCreate(K key, Func<T> creatorFunc)
+        public T FindOrCreate(K key, Func<K, T> creatorFunc)
         {
-            T obj = null;
-            WeakReference cacheEntry = internCache.GetOrAdd(key,
-                (k) =>
-                {
-                    obj = creatorFunc();
-                    return new WeakReference(obj);
-                });
-            if (cacheEntry != null)
+            T result;
+            WeakReference<T> cacheEntry;
+
+            // Attempt to get the existing value from cache.
+            internCache.TryGetValue(key, out cacheEntry);
+
+            // If no cache entry exists, create and insert a new one using the creator function.
+            if (cacheEntry == null)
```

Comment:
The `Func` passed to `GetOrAdd` are invoked outside of any lock. This effectively inlines the logic so that we do not need to allocate that closure.

This is the `GetOrAdd` code:
```C#
      TValue resultingValue;
      if (this.TryGetValue(key, out resultingValue))
        return resultingValue;
      this.TryAddInternal(key, valueFactory(key), updateIfExists: false, acquireLock: true, out resultingValue);
      return resultingValue;
```

Whereas the indexer setter is simply this:
```C#
TValue resultingValue;
this.TryAddInternal(key, value, updateIfExists: true, acquireLock: true, out resultingValue);
```

There's no point allocating the `valueFactory` closure (captures `creatorFunc`), it makes more sense (in my opinion) to just call `TryGetValue` and if the value doesn't exist or is no longer valid, then to create and set the value.

Additionally, the purpose of this class is to cache immutable values - if the values stored in this class do mutate, then that's an issue (i,.e, user is abusing this class).

Thanks for calling this out, by the way. I should have preemptively added a comment here to discuss my reasoning.